content/2015.12.loadtesting.with.oauth.rst

```diff
@@ -0,0 +1,171 @@
+How to load test a OAuth-protected HTTP API?
+############################################
+
+:lang: en
+:date: 2015-12-02
+:url: load-testing-a-http-api-which-uses-oauth-for-authentication
+:summary: This article explains how to test the performance of an HTTP API when its access depends on an external OAuth service.
+
+
+What is OAuth and how do you use it to authenticate an HTTP API?
+================================================================
+
+`OAuth <https://en.wikipedia.org/wiki/OAuth>`_ is a protocol that let
+a user delegate some of its permissions to a web service through an
+identity provider.
+
+The user can be confident that the web service will never have its
+user credentials (ID, password), instead it will get a Bearer Token
+that is valid for a list of scopes and for a certain amount of time.
+
+In the app, the user click on a Sign-In button that will redirect it
+to the Identity Provider login page with the list of scopes the app
+needs.
+
+The user will enter her login credentials and validate the permission
+list, once the credentials are correct, the Identity provider will
+redirect the user to the app with the
+`Bearer tokens <http://tools.ietf.org/html/rfc6750>`_ enclosed in the query string.
+
+.. image:: {filename}/images/oauth-flow.png
+   :alt: OAuth 2.0 Bearer Token creation flow
+
+This Bearer Token will then be used by the app in the HTTP request
+Authorization header to authenticate on the web service.
+
+.. code-block:: http
+
+    GET /v1/articles HTTP/1.1
+    Authorization: Bearer e7613cae7a0660dae7a2dd55c216cbaef75f43254e2a8c77f7e9c7adb549bb65
+    Host: readinglist.services.mozilla.com
+
+
+Why would you need to run a load test?
+======================================
+
+After deploying an HTTP API you might be wondering how many users or
+requests your deployment can take and what happens after it reaches
+that number.
+
+Knowing the maximum load your service can handle is a great thing to
+know!  It tells you when to start scaling up when your production
+server is close to reach that number of requests per second.
+
+
+So, what's so complicated about OAuth?
+======================================
+
+Load testing a service that requires OAuth-based authentication means:
+
+* having an OAuth account (often creating it) 
+* getting a Bearer Token with the right scopes
```

Comment:
A related concern (which I don't know whether you're dealing with or not in your use-case) is whether you test load with a single user account, or with lots of different user accounts.  The difference can affect the performance of your app, and using many different accounts is harder but more realistic.